pkg/discovery/k8s_discovery_client.go

```diff
@@ -257,12 +275,20 @@ func (dc *K8sDiscoveryClient) discoverWithNewFramework(targetID string) ([]*prot
 	nodes := clusterSummary.NodeList
 	// Call cache cleanup
 	dc.config.probeConfig.NodeClient.CleanupCache(nodes)
+	// Stops scheduling dispatcher to assign sampling discovery tasks.
+	dc.samplingDispatcher.FinishSampling()
 
 	// Discover pods and create DTOs for nodes, namespaces, controllers, pods, containers, application.
-	// Collect the kubePod, kubeNamespace metrics, groups and kubeControllers from all the discovery workers
+	// Merge collected usage data samples from globalEntityMetricSink into the metric sink of each individual discovery worker.
+	// Collect the kubePod, kubeNamespace metrics, groups and kubeControllers from all the discovery workers.
 	taskCount := dc.dispatcher.Dispatch(nodes, clusterSummary)
 	result := dc.resultCollector.Collect(taskCount)
 
+	// Clear globalEntityMetricSink cache after collecting main discovery results
+	dc.globalEntityMetricSink.ClearCache()
+	// Reschedule dispatch sampling discovery tasks for newly discovered nodes
+	dc.samplingDispatcher.ScheduleDispatch(nodes)
```

Comment:
[Probably no action needed but worth taking note]: This seems to mean that the nodes are refreshed only when discovery is trigerred. So if a node leaves in between discoveries, or a new one joins in during that time, the sampling data would be incoherent. This will especially be exacerbated when the probe runs disconnected from server, and then eventually joins in, or a patchy network, where probe keeps running online/offline over a period of time.

Reply:
Thanks @irfanurrehman, this is a good point. As we discussed, it would be expensive to get all nodes every sampling interval from cluster. And I'll keep the current logic, and add a comment and additional warn message if there's error when getting metrics summary from an unavailable node during sampling discovery.